File: cpp/src/market.cpp
#include "frtb/market.hpp"

#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <utility>

namespace frtb {

Curve::Curve(std::vector<double> tenors, std::vector<double> rates)
    : tenors_(std::move(tenors)), rates_(std::move(rates)) {
    if (tenors_.empty() || tenors_.size() != rates_.size())
        throw std::invalid_argument("Curve: tenors and rates must be non-empty and equal length");
    for (std::size_t i = 1; i < tenors_.size(); ++i)
        if (tenors_[i] <= tenors_[i - 1])
            throw std::invalid_argument("Curve: tenors must be strictly increasing");
    for (std::size_t i = 0; i < tenors_.size(); ++i)
        if (!(std::isfinite(tenors_[i]) && std::isfinite(rates_[i])) || tenors_[i] <= 0.0)
            throw std::invalid_argument("Curve: tenors must be positive finite, rates finite");
}

double Curve::rate(double t) const {
    if (!std::isfinite(t) || t < 0.0)
        throw std::invalid_argument("Curve::rate: invalid time " + std::to_string(t));
    const auto& ts = tenors_;
    const auto& rs = rates_;
    if (t <= ts.front()) return rs.front();
    if (t >= ts.back()) return rs.back();
    for (std::size_t i = 1; i < ts.size(); ++i) {
        if (t <= ts[i]) {
            // Same interpolation arithmetic as the Python reference.
            double w = (t - ts[i - 1]) / (ts[i] - ts[i - 1]);
            return rs[i - 1] * (1.0 - w) + rs[i] * w;
        }
    }
    return rs.back();  // unreachable
}

double Curve::df(double t) const {
    if (t == 0.0) return 1.0;
    return std::exp(-rate(t) * t);
}

Curve Curve::bumped_node(double tenor, double size) const {
    bool found = false;
    for (double tt : tenors_)
        if (tt == tenor) found = true;
    if (!found)
        throw std::invalid_argument("Curve::bumped_node: tenor " + std::to_string(tenor) +
                                    " is not a curve node");
    std::vector<double> rates = rates_;
    for (std::size_t i = 0; i < tenors_.size(); ++i)
        if (tenors_[i] == tenor) rates[i] = rates_[i] + size;
    return Curve(tenors_, rates);
}

Curve Curve::bumped_parallel(double size) const {
    std::vector<double> rates = rates_;
    for (double& r : rates) r += size;
    return Curve(tenors_, rates);
}

EquityQuote::EquityQuote(double spot_, double vol_, double div_yield_, std::string bucket_)
    : spot(spot_), vol(vol_), div_yield(div_yield_), bucket(std::move(bucket_)) {
    if (!std::isfinite(spot) || spot <= 0.0)
        throw std::invalid_argument("EquityQuote: spot must be positive finite");
    if (!std::isfinite(vol) || vol < 0.0)
        throw std::invalid_argument("EquityQuote: vol must be >= 0");
    if (!std::isfinite(div_yield))
        throw std::invalid_argument("EquityQuote: div_yield must be finite");
}

const Curve& Market::curve(const std::string& ccy) const {
    auto it = curves.find(ccy);
    if (it == curves.end())
        throw std::invalid_argument("Market: no curve for currency '" + ccy + "'");
    return it->second;
}

const EquityQuote& Market::equity(const std::string& name) const {
    auto it = equities.find(name);
    if (it == equities.end())
        throw std::invalid_argument("Market: no equity quote for '" + name + "'");
    return it->second;
}

double Market::fx_spot(const std::string& pair) const {
    auto it = fx.find(pair);
    if (it == fx.end())
        throw std::invalid_argument("Market: no FX spot for pair '" + pair + "'");
    return it->second;
}

Market Market::with_curve(const std::string& ccy, Curve curve) const {
    Market m = *this;
    m.curves.erase(ccy);
    m.curves.emplace(ccy, std::move(curve));
    return m;
}

Market Market::bump_curve_node(const std::string& ccy, double tenor, double size) const {
    return with_curve(ccy, curve(ccy).bumped_node(tenor, size));
}

Market Market::bump_curve_parallel(const std::string& ccy, double size) const {
    return with_curve(ccy, curve(ccy).bumped_parallel(size));
}

Market Market::bump_equity_spot(const std::string& name, double rel) const {
    const EquityQuote& q = equity(name);
    Market m = *this;
    m.equities[name] = EquityQuote(q.spot * (1.0 + rel), q.vol, q.div_yield, q.bucket);
    return m;
}

Market Market::bump_equity_vol(const std::string& name, double size) const {
    const EquityQuote& q = equity(name);
    Market m = *this;
    m.equities[name] = EquityQuote(q.spot, q.vol + size, q.div_yield, q.bucket);
    return m;
}

Market Market::bump_fx(const std::string& pair, double rel) const {
    double s = fx_spot(pair);
    Market m = *this;
    m.fx[pair] = s * (1.0 + rel);
    return m;
}

namespace {

/// Split one CSV line into cells (the bundled files have no quoting).
std::vector<std::string> split_csv(const std::string& line) {
    std::vector<std::string> out;
    std::string cell;
    std::istringstream ss(line);
    while (std::getline(ss, cell, ',')) out.push_back(cell);
    if (!line.empty() && line.back() == ',') out.emplace_back();
    return out;
}

std::string strip(const std::string& s) {
    std::size_t a = s.find_first_not_of(" \t\r\n");
    if (a == std::string::npos) return "";
    std::size_t b = s.find_last_not_of(" \t\r\n");
    return s.substr(a, b - a + 1);
}

double to_double(const std::string& s, const std::string& ctx) {
    char* end = nullptr;
    std::string t = strip(s);
    double v = std::strtod(t.c_str(), &end);
    if (t.empty() || end != t.c_str() + t.size())
        throw std::invalid_argument("load_market: bad number '" + s + "' in " + ctx);
    return v;
}

}  // namespace

Market load_market(const std::string& curves_csv, const std::string& spots_csv) {
    std::ifstream cf(curves_csv);
    if (!cf) throw std::invalid_argument("load_market: cannot open " + curves_csv);
    std::string line;
    if (!std::getline(cf, line))
        throw std::invalid_argument("load_market: empty file " + curves_csv);
    std::map<std::string, std::vector<std::pair<double, double>>> by_ccy;
    while (std::getline(cf, line)) {
        if (strip(line).empty()) continue;
        auto cells = split_csv(line);
        if (cells.size() != 3)
            throw std::invalid_argument("load_market: bad row in " + curves_csv);
        by_ccy[strip(cells[0])].emplace_back(to_double(cells[1], curves_csv),
                                             to_double(cells[2], curves_csv));
    }
    if (by_ccy.empty())
        throw std::invalid_argument("load_market: no curves in " + curves_csv);

    Market m;
    for (auto& [ccy, pts] : by_ccy) {
        std::sort(pts.begin(), pts.end());
        std::vector<double> tenors, rates;
        for (const auto& p : pts) {
            tenors.push_back(p.first);
            rates.push_back(p.second);
        }
        m.curves.emplace(ccy, Curve(tenors, rates));
    }

    std::ifstream sf(spots_csv);
    if (!sf) throw std::invalid_argument("load_market: cannot open " + spots_csv);
    if (!std::getline(sf, line))
        throw std::invalid_argument("load_market: empty file " + spots_csv);
    while (std::getline(sf, line)) {
        if (strip(line).empty()) continue;
        auto cells = split_csv(line);
        if (cells.size() < 6)
            throw std::invalid_argument("load_market: bad row in " + spots_csv);
        std::string kind = strip(cells[0]);
        std::string name = strip(cells[1]);
        if (kind == "equity") {
            m.equities[name] = EquityQuote(to_double(cells[2], spots_csv),
                                           to_double(cells[3], spots_csv),
                                           to_double(cells[4], spots_csv), strip(cells[5]));
        } else if (kind == "fx") {
            m.fx[name] = to_double(cells[2], spots_csv);
        } else {
            throw std::invalid_argument("load_market: unknown kind '" + kind + "' in " + spots_csv);
        }
    }
    return m;
}

}  // namespace frtb

File: rust/src/market.rs
//! Market data containers: zero curves, equity quotes, FX spots.
//!
//! All bump operations return *new* objects (the engine treats markets as
//! immutable snapshots), so bump-and-revalue sensitivities cannot leak
//! state between bumps.

use std::collections::BTreeMap;
use std::path::Path;

use crate::error::{invalid, FrtbError, Result};

/// Continuously-compounded zero curve with linear interpolation in tenor.
///
/// Rates are interpolated linearly between nodes and extrapolated flat
/// beyond the first/last node. Discount factor: `DF(t) = exp(-z(t)*t)`,
/// `DF(0) = 1`.
#[derive(Debug, Clone, PartialEq)]
pub struct Curve {
    tenors: Vec<f64>,
    rates: Vec<f64>,
}

impl Curve {
    /// Build a curve; tenors must be positive, finite and strictly
    /// increasing, rates finite.
    pub fn new(tenors: Vec<f64>, rates: Vec<f64>) -> Result<Curve> {
        if tenors.is_empty() || tenors.len() != rates.len() {
            return invalid("Curve: tenors and rates must be non-empty and equal length");
        }
        for i in 1..tenors.len() {
            if tenors[i] <= tenors[i - 1] {
                return invalid("Curve: tenors must be strictly increasing");
            }
        }
        for (&t, &r) in tenors.iter().zip(rates.iter()) {
            if !t.is_finite() || !r.is_finite() || t <= 0.0 {
                return invalid("Curve: tenors must be positive finite, rates finite");
            }
        }
        Ok(Curve { tenors, rates })
    }

    /// Curve node tenors (strictly increasing).
    pub fn tenors(&self) -> &[f64] {
        &self.tenors
    }

    /// Zero rates aligned with [`Curve::tenors`].
    pub fn rates(&self) -> &[f64] {
        &self.rates
    }

    /// Interpolated zero rate at time `t` (flat extrapolation).
    pub fn rate(&self, t: f64) -> Result<f64> {
        if !t.is_finite() || t < 0.0 {
            return invalid(format!("Curve.rate: invalid time {t}"));
        }
        let (ts, rs) = (&self.tenors, &self.rates);
        if t <= ts[0] {
            return Ok(rs[0]);
        }
        if t >= ts[ts.len() - 1] {
            return Ok(rs[rs.len() - 1]);
        }
        for i in 1..ts.len() {
            if t <= ts[i] {
                let w = (t - ts[i - 1]) / (ts[i] - ts[i - 1]);
                return Ok(rs[i - 1] * (1.0 - w) + rs[i] * w);
            }
        }
        Ok(rs[rs.len() - 1]) // unreachable: t < last tenor was handled above
    }

    /// Discount factor `exp(-z(t)*t)`; `DF(0) = 1`.
    pub fn df(&self, t: f64) -> Result<f64> {
        if t == 0.0 {
            return Ok(1.0);
        }
        Ok((-self.rate(t)? * t).exp())
    }

    /// A copy with the zero rate at one node shifted by `size` (absolute).
    ///
    /// Errors when `tenor` is not an exact curve node (the pinned GIRR bump
    /// convention bumps nodes, never interpolated points).
    pub fn bumped_node(&self, tenor: f64, size: f64) -> Result<Curve> {
        if !self.tenors.contains(&tenor) {
            return invalid(format!("Curve.bumped_node: tenor {tenor} is not a curve node"));
        }
        let rates = self
            .tenors
            .iter()
            .zip(self.rates.iter())
            .map(|(&tt, &r)| if tt == tenor { r + size } else { r })
            .collect();
        Curve::new(self.tenors.clone(), rates)
    }

    /// A copy with every node shifted by `size` (absolute).
    pub fn bumped_parallel(&self, size: f64) -> Result<Curve> {
        let rates = self.rates.iter().map(|&r| r + size).collect();
        Curve::new(self.tenors.clone(), rates)
    }
}

/// Equity market data: spot, flat lognormal vol, dividend yield, SBM bucket.
#[derive(Debug, Clone, PartialEq)]
pub struct EquityQuote {
    /// Spot price (positive, finite).
    pub spot: f64,
    /// Flat lognormal volatility (>= 0).
    pub vol: f64,
    /// Continuous dividend yield.
    pub div_yield: f64,
    /// SBM equity bucket label (e.g. "1", "11").
    pub bucket: String,
}

impl EquityQuote {
    /// Build a quote with input validation.
    pub fn new(spot: f64, vol: f64, div_yield: f64, bucket: String) -> Result<EquityQuote> {
        if !spot.is_finite() || spot <= 0.0 {
            return invalid(format!("EquityQuote: spot must be positive finite, got {spot}"));
        }
        if !vol.is_finite() || vol < 0.0 {
            return invalid(format!("EquityQuote: vol must be >= 0, got {vol}"));
        }
        if !div_yield.is_finite() {
            return invalid("EquityQuote: div_yield must be finite");
        }
        Ok(EquityQuote { spot, vol, div_yield, bucket })
    }
}

/// Immutable market snapshot: curves per currency, equities per name, FX
/// spots per pair. All `bump_*` helpers return new snapshots.
#[derive(Debug, Clone, PartialEq)]
pub struct Market {
    /// Zero curves keyed by currency code.
    pub curves: BTreeMap<String, Curve>,
    /// Equity quotes keyed by underlier name.
    pub equities: BTreeMap<String, EquityQuote>,
    /// FX spots keyed by pair (FORDOM, e.g. "EURUSD").
    pub fx: BTreeMap<String, f64>,
}

impl Market {
    /// Curve for `ccy` or an error naming the missing currency.
    pub fn curve(&self, ccy: &str) -> Result<&Curve> {
        self.curves
            .get(ccy)
            .ok_or_else(|| FrtbError::Invalid(format!("Market: no curve for currency '{ccy}'")))
    }

    /// Equity quote for `name` or an error.
    pub fn equity(&self, name: &str) -> Result<&EquityQuote> {
        self.equities
            .get(name)
            .ok_or_else(|| FrtbError::Invalid(format!("Market: no equity quote for '{name}'")))
    }

    /// FX spot for `pair` or an error.
    pub fn fx_spot(&self, pair: &str) -> Result<f64> {
        self.fx
            .get(pair)
            .copied()
            .ok_or_else(|| FrtbError::Invalid(format!("Market: no FX spot for pair '{pair}'")))
    }

    /// New snapshot with the curve for `ccy` replaced.
    pub fn with_curve(&self, ccy: &str, curve: Curve) -> Market {
        let mut curves = self.curves.clone();
        curves.insert(ccy.to_string(), curve);
        Market { curves, equities: self.equities.clone(), fx: self.fx.clone() }
    }

    /// Snapshot with one zero-curve node bumped by `size` (absolute).
    pub fn bump_curve_node(&self, ccy: &str, tenor: f64, size: f64) -> Result<Market> {
        Ok(self.with_curve(ccy, self.curve(ccy)?.bumped_node(tenor, size)?))
    }

    /// Snapshot with the whole curve of `ccy` bumped by `size` (absolute).
    pub fn bump_curve_parallel(&self, ccy: &str, size: f64) -> Result<Market> {
        Ok(self.with_curve(ccy, self.curve(ccy)?.bumped_parallel(size)?))
    }

    /// Snapshot with a relative equity-spot bump: `S -> S * (1 + rel)`.
    pub fn bump_equity_spot(&self, name: &str, rel: f64) -> Result<Market> {
        let q = self.equity(name)?;
        let bumped = EquityQuote::new(q.spot * (1.0 + rel), q.vol, q.div_yield, q.bucket.clone())?;
        let mut equities = self.equities.clone();
        equities.insert(name.to_string(), bumped);
        Ok(Market { curves: self.curves.clone(), equities, fx: self.fx.clone() })
    }

    /// Snapshot with an absolute vol bump: `sigma -> sigma + size`.
    pub fn bump_equity_vol(&self, name: &str, size: f64) -> Result<Market> {
        let q = self.equity(name)?;
        let bumped = EquityQuote::new(q.spot, q.vol + size, q.div_yield, q.bucket.clone())?;
        let mut equities = self.equities.clone();
        equities.insert(name.to_string(), bumped);
        Ok(Market { curves: self.curves.clone(), equities, fx: self.fx.clone() })
    }

    /// Snapshot with a relative FX-spot bump: `S -> S * (1 + rel)`.
    pub fn bump_fx(&self, pair: &str, rel: f64) -> Result<Market> {
        let s = self.fx_spot(pair)?;
        let mut fx = self.fx.clone();
        fx.insert(pair.to_string(), s * (1.0 + rel));
        Ok(Market { curves: self.curves.clone(), equities: self.equities.clone(), fx })
    }
}

// ---------------------------------------------------------------- loading

fn read_csv(path: &Path) -> Result<(Vec<String>, Vec<Vec<String>>)> {
    let text = std::fs::read_to_string(path)
        .map_err(|e| FrtbError::Io(format!("cannot read {}: {e}", path.display())))?;
    let mut lines = text.lines().filter(|l| !l.trim().is_empty());
    let header: Vec<String> = lines
        .next()
        .ok_or_else(|| FrtbError::Io(format!("{}: empty CSV", path.display())))?
        .split(',')
        .map(|s| s.trim().to_string())
        .collect();
    let rows = lines
        .map(|l| l.split(',').map(|s| s.trim().to_string()).collect())
        .collect();
    Ok((header, rows))
}

fn col(header: &[String], name: &str, path: &Path) -> Result<usize> {
    header
        .iter()
        .position(|h| h == name)
        .ok_or_else(|| FrtbError::Io(format!("{}: missing column '{name}'", path.display())))
}

fn parse_f64(cell: &str, ctx: &str) -> Result<f64> {
    cell.parse::<f64>()
        .map_err(|_| FrtbError::Invalid(format!("{ctx}: cannot parse number '{cell}'")))
}

/// Load a [`Market`] from `curves.csv` (`currency,tenor,zero_rate`) and
/// `spots.csv` (`kind,name,spot,vol,div_yield,eq_bucket`).
pub fn load_market(curves_csv: &Path, spots_csv: &Path) -> Result<Market> {
    let (header, rows) = read_csv(curves_csv)?;
    let (c_ccy, c_tenor, c_rate) = (
        col(&header, "currency", curves_csv)?,
        col(&header, "tenor", curves_csv)?,
        col(&header, "zero_rate", curves_csv)?,
    );
    let mut by_ccy: BTreeMap<String, Vec<(f64, f64)>> = BTreeMap::new();
    for row in &rows {
        by_ccy.entry(row[c_ccy].clone()).or_default().push((
            parse_f64(&row[c_tenor], "curves.csv tenor")?,
            parse_f64(&row[c_rate], "curves.csv zero_rate")?,
        ));
    }
    if by_ccy.is_empty() {
        return invalid(format!("load_market: no curves in {}", curves_csv.display()));
    }
    let mut curves = BTreeMap::new();
    for (ccy, mut pts) in by_ccy {
        pts.sort_by(|a, b| a.partial_cmp(b).expect("finite tenors"));
        let tenors = pts.iter().map(|p| p.0).collect();
        let rates = pts.iter().map(|p| p.1).collect();
        curves.insert(ccy, Curve::new(tenors, rates)?);
    }

    let (header, rows) = read_csv(spots_csv)?;
    let (c_kind, c_name, c_spot, c_vol, c_div, c_bkt) = (
        col(&header, "kind", spots_csv)?,
        col(&header, "name", spots_csv)?,
        col(&header, "spot", spots_csv)?,
        col(&header, "vol", spots_csv)?,
        col(&header, "div_yield", spots_csv)?,
        col(&header, "eq_bucket", spots_csv)?,
    );
    let mut equities = BTreeMap::new();
    let mut fx = BTreeMap::new();
    for row in &rows {
        match row[c_kind].as_str() {
            "equity" => {
                let q = EquityQuote::new(
                    parse_f64(&row[c_spot], "spots.csv spot")?,
                    parse_f64(&row[c_vol], "spots.csv vol")?,
                    parse_f64(&row[c_div], "spots.csv div_yield")?,
                    row[c_bkt].clone(),
                )?;
                equities.insert(row[c_name].clone(), q);
            }
            "fx" => {
                fx.insert(row[c_name].clone(), parse_f64(&row[c_spot], "spots.csv spot")?);
            }
            other => {
                return invalid(format!(
                    "load_market: unknown kind '{other}' in {}",
                    spots_csv.display()
                ));
            }
        }
    }
    Ok(Market { curves, equities, fx })
}
